### PR TITLE
2960-user-specific-task-url

### PR DESCRIPTION
**Pull request by afs25**
Ready for review.

### CODE REVIEW

File: api/endpoints/user_task.py [hunk 1]
```diff
@@ -81,7 +82,9 @@ def list_user_tasks_by_user(user_id, correlation_id=None):
     edited_result = list()
     for ut in result:
         base_url = ut['base_url']
-        if base_url is not None:
+        if ut['user_specific_url']:
+            url = ut['user_task_url']
+        elif base_url is not None:
             user_id = ut['user_id']
             external_task_id = ut['external_task_id']
             url = ut['base_url'] + utils.create_url_params(user_id, ut['user_task_id'], external_task_id) + utils.non_prod_env_url_param()
```
Comment (this-andy) [inline api/endpoints/user_task.py:90]: Move addition of params outside if stmt

File: api/endpoints/project.py [hunk 2]
```diff
@@ -132,7 +132,12 @@ def get_project_status_for_user(user_id, correlation_id, anonymise_url=False):
                         task['user_task_status'] = projects_usertasks_dict[task_id]['status']
                 # only give url if user has signedup (inc if completed)
                 if task['task_is_visible'] and task['user_is_signedup']:
-                    if task['url'] is not None:
+                    if task['user_specific_url']:
+                        task['url'] = "{}{}".format(
+                            projects_usertasks_dict[task_id]['user_task_url'],
+                            utils.non_prod_env_url_param()
+                        )
```
Comment (this-andy) [inline api/endpoints/project.py:139]: Need to add params to user specific url

File: api/endpoints/user_task.py [hunk 3]
```diff
@@ -201,9 +210,21 @@ def create_user_task(ut_json, correlation_id):
...
 
-    execute_non_query(CREATE_USER_TASK_SQL, (id, created, created, user_project_id, project_task_id, status, consented, ext_user_task_id), correlation_id)
+    execute_non_query(
+        CREATE_USER_TASK_SQL,
+        (id, created, created, user_project_id, project_task_id, status, consented, ext_user_task_id, user_task_url),
+        correlation_id
+    )
+
+    if user_specific_url:
+        raw_url = user_task_url
```
Comment (this-andy) [inline api/endpoints/user_task.py:220]: params again

**Comment by afs25** [comment]
Made requested changes (and added more tests), so this is ready for review again.

File: api/endpoints/user_task.py [hunk 4]
```diff
@@ -162,7 +163,7 @@ def create_user_task(ut_json, correlation_id):
         ('ext_user_task_id', str(uuid.uuid4()), utils.validate_uuid),
         ('created', str(utils.now_with_tz()), utils.validate_utc_datetime),
         ('status', DEFAULT_STATUS, validate_status),
-        ('user_task_url', None, validators.url)
+        ('user_task_url', None, utils.validate_url)
```
Comment (this-andy) [inline api/endpoints/user_task.py:166]: Can we remove this param

**Pull request merged**